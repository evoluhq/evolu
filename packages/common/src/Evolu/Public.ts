/**
 * ðŸ’¾
 *
 * @module
 */

export { createEvolu } from "./Evolu.js";
export type { Evolu, EvoluDeps, EvoluError } from "./Evolu.js";
export * from "./Owner.js";
export { binaryIdToId, idToBinaryId } from "./Protocol.js";
export type { BinaryId } from "./Protocol.js";
export * as kysely from "./PublicKysely.js";
export type { InferRow, Query, QueryRows, Row } from "./Query.js";
export { formatValidMutationSizeError } from "./Schema.js";
export type { EvoluSchema, ValidMutationSizeError } from "./Schema.js";
export type {
  NetworkError,
  PaymentRequiredError,
  ServerError,
  SyncState,
  SyncStateInitial,
  SyncStateIsNotSynced,
  SyncStateIsSynced,
  SyncStateIsSyncing,
} from "./Sync.js";
export {
  binaryTimestampToTimestamp,
  Timestamp,
  timestampToBinaryTimestamp,
} from "./Timestamp.js";
export type {
  BinaryTimestamp,
  TimestampCounterOverflowError,
  TimestampDriftError,
  TimestampDuplicateNodeError,
  TimestampError,
  TimestampTimeOutOfRangeError,
} from "./Timestamp.js";
