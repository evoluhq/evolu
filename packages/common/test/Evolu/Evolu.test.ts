import { describe, expect, expectTypeOf, test, vi } from "vitest";
import { createConsole } from "../../src/Console.js";
import { createEvolu } from "../../src/Evolu/Evolu.js";
import { createAppOwner } from "../../src/Evolu/Owner.js";
import { getOrThrow } from "../../src/Result.js";
import { createSqlite, SqliteBoolean } from "../../src/Sqlite.js";
import {
  Boolean,
  id,
  InferType,
  maxLength,
  NonEmptyString,
  nullOr,
  SimpleName,
} from "../../src/Type.js";
import {
  testCreateId,
  testCreateMnemonic,
  testCreateRandomBytesDep,
  testCreateSqliteDriver,
  testMnemonic,
  testNanoIdLib,
  testRandom,
  testSimpleName,
  testTime,
} from "../_deps.js";
import {
  ValidateColumnTypes,
  ValidateIdColumnType,
  ValidateNoDefaultColumns,
  ValidateSchemaHasId,
} from "../../src/Evolu/Schema.js";
import {
  createDbWorkerForPlatform,
  getDbSnapshot,
} from "../../src/Evolu/Db.js";
import { constVoid } from "../../src/Function.js";
import { wait } from "../../src/Promise.js";

const TodoId = id("Todo");
type TodoId = InferType<typeof TodoId>;

const TodoCategoryId = id("TodoCategory");
type TodoCategoryId = InferType<typeof TodoCategoryId>;

const NonEmptyString50 = maxLength(50)(NonEmptyString);
type NonEmptyString50 = InferType<typeof NonEmptyString50>;

const Schema = {
  todo: {
    id: TodoId,
    title: NonEmptyString50,
    isCompleted: nullOr(SqliteBoolean),
    categoryId: nullOr(TodoCategoryId),
  },
  todoCategory: {
    id: TodoCategoryId,
    name: NonEmptyString50,
  },
};

const mockDeps = () => {
  const dbWorker = {
    onMessage: vi.fn(),
    postMessage: vi.fn(),
  };

  return {
    createDbWorker: () => dbWorker,
    time: testTime,
    nanoIdLib: testNanoIdLib,
    console: createConsole(),
    createAppState: () => ({
      reset: vi.fn(),
    }),
  };
};

let instancesCount = 0;

const setupEvoluTest = () => {
  const deps = mockDeps();
  const evolu = createEvolu(deps)(Schema, {
    name: getOrThrow(SimpleName.from(`instance${instancesCount++}`)),
  });
  const dbWorker = deps.createDbWorker();

  return { deps, evolu, dbWorker };
};

const createEvoluDepsWithSqlite = async () => {
  const sqliteDriver = testCreateSqliteDriver(testSimpleName);

  const dbWorker = createDbWorkerForPlatform({
    createSqliteDriver: () => sqliteDriver,
    createSync: () => () => ({ send: constVoid }),
    console: createConsole(),
    time: testTime,
    random: testRandom,
    nanoIdLib: testNanoIdLib,
    createMnemonic: testCreateMnemonic,
    createRandomBytes: testCreateRandomBytesDep.createRandomBytes,
  });

  const deps = {
    createDbWorker: () => dbWorker,
    time: testTime,
    nanoIdLib: testNanoIdLib,
    console: createConsole(),
    createAppState: () => ({ reset: vi.fn() }),
  };

  const sqlite = getOrThrow(
    await createSqlite({
      createSqliteDriver: () => sqliteDriver,
    })(testSimpleName),
  );

  return { deps, sqliteDriver, sqlite };
};

test("init postMessage call", () => {
  const { dbWorker } = setupEvoluTest();

  expect(dbWorker.postMessage.mock.calls).toMatchInlineSnapshot(`
    [
      [
        {
          "config": {
            "enableLogging": false,
            "maxDrift": 300000,
            "name": "instance0",
            "reloadUrl": "/",
            "syncUrl": "wss://free.evoluhq.com",
          },
          "dbSchema": {
            "indexes": [],
            "tables": [
              {
                "columns": [
                  "title",
                  "isCompleted",
                  "categoryId",
                ],
                "name": "todo",
              },
              {
                "columns": [
                  "name",
                ],
                "name": "todoCategory",
              },
            ],
          },
          "type": "init",
        },
      ],
    ]
  `);
});

test("insert should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const validTodo = {
    title: "Test Todo",
  };

  const result = evolu.insert("todo", validTodo);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toMatchInlineSnapshot(
    `"3C22DRVU0AHGjXpOEP-WJ"`,
  );

  await wait(0);

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "3C22DRVU0AHGjXpOEP-WJ",
            "table": "todo",
            "values": {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "title": "Test Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "kYF3FmbitSesTwHwplqLB",
        "type": "mutate",
      },
    ]
  `);
});

test("insert should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  // Empty title (violates NonEmptyString constraint)
  const invalidTodo = {
    title: "",
  };

  const result = evolu.insert("todo", invalidTodo);

  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "title": {
              "min": 1,
              "type": "MinLength",
              "value": "",
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "title": "",
        },
      },
      "ok": false,
    }
  `);
});

test("update should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const testId = testCreateId();

  // Valid update
  const validUpdate = {
    id: testId,
    title: "Updated Todo",
  };

  const result = evolu.update("todo", validUpdate);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toBe(testId);

  await wait(0);

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "D2PtSrFu-SJV0Ui1_SJB3",
            "table": "todo",
            "values": {
              "title": "Updated Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "kYF3FmbitSesTwHwplqLB",
        "type": "mutate",
      },
    ]
  `);
});

test("update should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  // Missing id
  const invalidUpdate = {
    title: "Updated Todo",
  };

  // @ts-expect-error - Testing runtime validation
  const result = evolu.update("todo", invalidUpdate);

  // Should return error
  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "id": {
              "type": "String",
              "value": undefined,
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "title": "Updated Todo",
        },
      },
      "ok": false,
    }
  `);
});

test("upsert should validate input and call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  const testId = testCreateId();

  // Valid upsert
  const validUpsert = {
    id: testId,
    title: "Upserted Todo",
  };

  const result = evolu.upsert("todo", validUpsert);

  expect(result.ok).toBe(true);
  expect(result.ok && result.value.id).toBe(testId);

  await wait(0);

  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "v5rPltodHge37rn9q4lwi",
            "table": "todo",
            "values": {
              "createdAt": "1970-01-01T00:00:00.001Z",
              "title": "Upserted Todo",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "kYF3FmbitSesTwHwplqLB",
        "type": "mutate",
      },
    ]
  `);
});

test("upsert should reject invalid input", () => {
  const { evolu } = setupEvoluTest();

  const testId = testCreateId();

  const invalidUpsert = {
    id: testId,
    title: "",
  };

  const result = evolu.upsert("todo", invalidUpsert);

  expect(result).toMatchInlineSnapshot(`
    {
      "error": {
        "reason": {
          "errors": {
            "title": {
              "min": 1,
              "type": "MinLength",
              "value": "",
            },
          },
          "kind": "Props",
        },
        "type": "Object",
        "value": {
          "id": "6xwHpK2ZkuZUN-T4MZhx0",
          "title": "",
        },
      },
      "ok": false,
    }
  `);
});

test("mutations should be processed in microtask queue", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  // Queue multiple mutations
  evolu.insert("todo", { title: "Todo 1" });
  evolu.insert("todo", { title: "Todo 2" });
  evolu.insert("todo", { title: "Todo 3" });

  await wait(0);

  // Only one postMessage call should happen with all changes
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(2); // 1 for init, 1 for mutations
  expect(dbWorker.postMessage.mock.calls[1]).toMatchInlineSnapshot(`
    [
      {
        "changes": [
          {
            "id": "yVRcpEppKHyKrUtl5RZ4m",
            "table": "todo",
            "values": {
              "createdAt": "1970-01-01T00:00:00.002Z",
              "title": "Todo 1",
            },
          },
          {
            "id": "CTxiAw0gY_fIT5Ci6Vt_f",
            "table": "todo",
            "values": {
              "createdAt": "1970-01-01T00:00:00.002Z",
              "title": "Todo 2",
            },
          },
          {
            "id": "kbP-KUG7NKoSfTGGJoCBJ",
            "table": "todo",
            "values": {
              "createdAt": "1970-01-01T00:00:00.002Z",
              "title": "Todo 3",
            },
          },
        ],
        "onCompleteIds": [],
        "subscribedQueries": [],
        "tabId": "kYF3FmbitSesTwHwplqLB",
        "type": "mutate",
      },
    ]
  `);
});

test("mutation with onlyValidate should not call postMessage", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  evolu.insert("todo", { title: "Validation only" }, { onlyValidate: true });

  await wait(0);

  // Only init should be called, not the mutation
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(1);
});

test("mutations should fail as a transaction when any mutation fails", async () => {
  const { evolu, dbWorker } = setupEvoluTest();

  // Queue valid and invalid mutations
  evolu.insert("todo", { title: "Valid Todo" });
  evolu.insert("todo", { title: "" }); // Invalid - empty title
  evolu.insert("todo", { title: "Another Valid Todo" });

  await wait(0);

  // Only init should be called, not the mutations since one failed
  expect(dbWorker.postMessage).toHaveBeenCalledTimes(1);
});

describe("EvoluSchema validation", () => {
  test("schema without id column", () => {
    const deps = mockDeps();

    const SchemaWithoutId = {
      todo: {
        // Missing id column - should cause TypeScript error
        title: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateSchemaHasId<typeof SchemaWithoutId>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" is missing required id column.'>();

    // @ts-expect-error - Schema validation should catch missing id column
    createEvolu(deps)(SchemaWithoutId, {
      name: testSimpleName,
    });
  });

  test("schema with default column createdAt", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        createdAt: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "createdAt". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with default column updatedAt", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        updatedAt: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "updatedAt". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with default column isDeleted", () => {
    const deps = mockDeps();

    const SchemaWithDefaultColumn = {
      todo: {
        id: TodoId,
        isDeleted: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateNoDefaultColumns<
      typeof SchemaWithDefaultColumn
    >;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" uses default column name "isDeleted". Default columns (createdAt, updatedAt, isDeleted) are added automatically.'>();

    // @ts-expect-error - Schema validation should catch default column name
    createEvolu(deps)(SchemaWithDefaultColumn, {
      name: testSimpleName,
    });
  });

  test("schema with non-branded id column", () => {
    const deps = mockDeps();

    const SchemaWithInvalidId = {
      todo: {
        id: NonEmptyString50,
        title: NonEmptyString50,
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateIdColumnType<typeof SchemaWithInvalidId>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" id column must be a branded ID type (created with id("todo")).'>();

    // @ts-expect-error - Schema validation should catch non-branded id column
    createEvolu(deps)(SchemaWithInvalidId, {
      name: testSimpleName,
    });
  });

  test("schema with incompatible column type", () => {
    const deps = mockDeps();

    const SchemaWithInvalidType = {
      todo: {
        id: TodoId,
        title: NonEmptyString50,
        invalidColumn: Boolean, // Boolean is not compatible with SQLite
      },
    };

    // Type-level assertion for the exact error message
    type ValidationResult = ValidateColumnTypes<typeof SchemaWithInvalidType>;
    expectTypeOf<ValidationResult>().toEqualTypeOf<'❌ Schema Error: Table "todo" column "invalidColumn" type is not compatible with SQLite. Column types must extend SqliteValue (string, number, Uint8Array, or null).'>();

    // @ts-expect-error - Schema validation should catch incompatible column type
    createEvolu(deps)(SchemaWithInvalidType, {
      name: testSimpleName,
    });
  });
});

describe("createdAt behavior", () => {
  test("insert should set createdAt to current time", async () => {
    const { evolu, dbWorker } = setupEvoluTest();

    const result = evolu.insert("todo", { title: "Test Todo" });
    expect(result.ok).toBe(true);

    await wait(0);

    // Verify the postMessage was called with createdAt in the change values
    expect(dbWorker.postMessage).toHaveBeenCalledWith(
      expect.objectContaining({
        type: "mutate",
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        changes: expect.arrayContaining([
          expect.objectContaining({
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            values: expect.objectContaining({
              title: "Test Todo",
              // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
              createdAt: expect.stringMatching(
                /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/,
              ),
            }),
          }),
        ]),
      }),
    );
  });

  test("upsert should set createdAt to current time", async () => {
    const { evolu, dbWorker } = setupEvoluTest();

    const testId = testCreateId();
    const result = evolu.upsert("todo", { id: testId, title: "Upserted Todo" });
    expect(result.ok).toBe(true);

    await wait(0);

    expect(dbWorker.postMessage).toHaveBeenCalledWith(
      expect.objectContaining({
        type: "mutate",
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        changes: expect.arrayContaining([
          expect.objectContaining({
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
            values: expect.objectContaining({
              title: "Upserted Todo",
              // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
              createdAt: expect.stringMatching(
                /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/,
              ),
            }),
          }),
        ]),
      }),
    );
  });

  test("update should NOT set createdAt", async () => {
    const { evolu, dbWorker } = setupEvoluTest();

    const testId = testCreateId();
    const result = evolu.update("todo", { id: testId, title: "Updated Todo" });
    expect(result.ok).toBe(true);

    await wait(0);

    // Get the actual call to inspect the values
    const calls = dbWorker.postMessage.mock.calls;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const mutateCall = calls.find((call) => call[0]?.type === "mutate");
    expect(mutateCall).toBeDefined();

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
    const change = mutateCall?.[0]?.changes?.[0];
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(change?.values).toEqual({
      title: "Updated Todo",
    });
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(change?.values).not.toHaveProperty("createdAt");
  });
});

test("initialAppOwner should use provided owner", async () => {
  const { deps, sqlite } = await createEvoluDepsWithSqlite();

  const initialAppOwner = createAppOwner(testMnemonic);

  createEvolu(deps)(Schema, {
    name: getOrThrow(SimpleName.from(`instance${instancesCount++}`)),
    initialAppOwner,
  });

  await wait(10);

  const snapshot = getDbSnapshot({ sqlite });
  expect(snapshot).toMatchSnapshot();

  const configTable = snapshot.tables.find(
    (table) => table.name === "evolu_config",
  );
  expect(configTable?.rows[0].appOwnerId).toBe(initialAppOwner.id);
});

test("onInit callback should be called with correct parameters and can seed initial data", async () => {
  const { deps, sqlite } = await createEvoluDepsWithSqlite();

  const initialAppOwner = createAppOwner(testMnemonic);
  const initCalls: Array<{
    appOwner: typeof initialAppOwner;
    isFirst: boolean;
  }> = [];

  const name = getOrThrow(SimpleName.from(`instance${instancesCount++}`));

  const evolu1 = createEvolu(deps)(Schema, {
    initialAppOwner,
    name,
    onInit: ({ appOwner, isFirst }) => {
      initCalls.push({ appOwner, isFirst });

      if (isFirst) {
        const todoCategoryId = getOrThrow(
          evolu1.insert("todoCategory", {
            name: "Not Urgent",
          }),
        );

        evolu1.insert("todo", {
          title: "Try React Suspense",
          categoryId: todoCategoryId.id,
        });
      }
    },
  });

  await wait(10);

  expect(initCalls).toHaveLength(1);

  const snapshot = getDbSnapshot({ sqlite });
  expect(snapshot).toMatchSnapshot();

  // Create
  createEvolu(deps)(Schema, {
    name,
    initialAppOwner,
    onInit: ({ appOwner, isFirst }) => {
      initCalls.push({ appOwner, isFirst });
    },
  });

  await wait(10);

  expect(initCalls).toHaveLength(1);
});
