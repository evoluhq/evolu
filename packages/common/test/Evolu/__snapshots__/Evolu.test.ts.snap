// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`initialAppOwner should use provided owner 1`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "title",
          "isCompleted",
          "categoryId",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todo",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todoCategory",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              173,
              165,
              204,
              225,
              74,
              240,
              76,
              99,
              46,
              18,
              104,
              4,
              0,
              136,
              188,
              132,
              99,
              57,
              151,
              4,
              147,
              44,
              59,
              91,
              61,
              162,
              116,
              243,
              186,
              98,
              25,
              4,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
          "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
          "appOwnerWriteKey": {
            "data": [
              111,
              64,
              118,
              108,
              245,
              65,
              176,
              98,
              173,
              247,
              85,
              78,
              30,
              167,
              192,
              203,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.000Z-0000-dea8e2c5a2d7f4c3",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [],
    },
    {
      "name": "todo",
      "rows": [],
    },
    {
      "name": "todoCategory",
      "rows": [],
    },
    {
      "name": "evolu_timestamp",
      "rows": [],
    },
  ],
}
`;

exports[`onInit callback should be called with correct parameters and can seed initial data 1`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "title",
          "isCompleted",
          "categoryId",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todo",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "todoCategory",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              173,
              165,
              204,
              225,
              74,
              240,
              76,
              99,
              46,
              18,
              104,
              4,
              0,
              136,
              188,
              132,
              99,
              57,
              151,
              4,
              147,
              44,
              59,
              91,
              61,
              162,
              116,
              243,
              186,
              98,
              25,
              4,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "MdVYFAxShUluuZKVWQfYL",
          "appOwnerMnemonic": "knee easy fork attitude drink gloom head latin spider grab spy reason",
          "appOwnerWriteKey": {
            "data": [
              111,
              64,
              118,
              108,
              245,
              65,
              176,
              98,
              173,
              247,
              85,
              78,
              30,
              167,
              192,
              203,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.011Z-0001-79bdc6f641541756",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [
        {
          "column": "name",
          "id": {
            "data": [
              45,
              57,
              23,
              185,
              183,
              25,
              242,
              209,
              61,
              55,
              167,
              255,
              131,
              135,
              18,
              120,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "table": "todoCategory",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
          "value": "Not Urgent",
        },
        {
          "column": "createdAt",
          "id": {
            "data": [
              45,
              57,
              23,
              185,
              183,
              25,
              242,
              209,
              61,
              55,
              167,
              255,
              131,
              135,
              18,
              120,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "table": "todoCategory",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
          "value": "1970-01-01T00:00:00.009Z",
        },
        {
          "column": "title",
          "id": {
            "data": [
              82,
              251,
              255,
              80,
              65,
              47,
              92,
              248,
              204,
              101,
              188,
              59,
              0,
              238,
              77,
              140,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
          "value": "Try React Suspense",
        },
        {
          "column": "categoryId",
          "id": {
            "data": [
              82,
              251,
              255,
              80,
              65,
              47,
              92,
              248,
              204,
              101,
              188,
              59,
              0,
              238,
              77,
              140,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
          "value": "TXBxQKVArGni9yYtMvVPR",
        },
        {
          "column": "createdAt",
          "id": {
            "data": [
              82,
              251,
              255,
              80,
              65,
              47,
              92,
              248,
              204,
              101,
              188,
              59,
              0,
              238,
              77,
              140,
            ],
            "type": "Buffer",
          },
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "table": "todo",
          "timestamp": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
          "value": "1970-01-01T00:00:00.009Z",
        },
      ],
    },
    {
      "name": "todo",
      "rows": [
        {
          "categoryId": "TXBxQKVArGni9yYtMvVPR",
          "createdAt": "1970-01-01T00:00:00.009Z",
          "id": "7ZZt7nnZx3D1AKbzu8w9D",
          "isCompleted": null,
          "isDeleted": null,
          "title": "Try React Suspense",
          "updatedAt": "1970-01-01T00:00:00.011Z",
        },
      ],
    },
    {
      "name": "todoCategory",
      "rows": [
        {
          "createdAt": "1970-01-01T00:00:00.009Z",
          "id": "TXBxQKVArGni9yYtMvVPR",
          "isDeleted": null,
          "name": "Not Urgent",
          "updatedAt": "1970-01-01T00:00:00.011Z",
        },
      ],
    },
    {
      "name": "evolu_timestamp",
      "rows": [
        {
          "c": 1,
          "h1": 275690730879081,
          "h2": 112511489795540,
          "l": 2,
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "t": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              0,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
        },
        {
          "c": 1,
          "h1": 103802921224654,
          "h2": 6929174574082,
          "l": 1,
          "ownerId": {
            "data": [
              128,
              87,
              31,
              173,
              149,
              230,
              206,
              93,
              128,
              2,
              246,
              220,
              162,
              236,
              95,
              168,
            ],
            "type": "Buffer",
          },
          "t": {
            "data": [
              0,
              0,
              0,
              0,
              0,
              11,
              0,
              1,
              121,
              189,
              198,
              246,
              65,
              84,
              23,
              86,
            ],
            "type": "Buffer",
          },
        },
      ],
    },
  ],
}
`;
